### ISSUE TITLE
TestEquality clarification from core library devs

### ISSUE THREAD

**Issue by robdockins** [GaloisInc/parameterized-utils#129]
The following patch has been approved and should eventually make it's way into `base`.

https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7333/diffs

It clarifies that `testEquality` should return `Just Refl` _iff_ the corresponding type indices are the same, and reiterates that it is desirable that the class only be inhabited by singletons.  This is incompatible with many of our typical uses of this class.
 
The actual class declaration will not change, so nothing will immediately break with future releases of base. However, going forward we will have to decide what to do about the fact that many of our uses of `TestEquality` will be considered law-breaking.

**Comment by travitch** [comment]
Fun. Since nearly all of our uses import the required class via `Data.Parameterized.Classes` (instead of the type equality module in base), we could always just duplicate the class definition with dire warnings that they are not the same. That would save us from having to change much client code.

That might ultimately be bad and confusing, though.

**Comment by robdockins** [comment]
I think duplicating the class and reusing the name would be strictly worse than doing nothing and having law-breaking instances.

**Comment by benjaminselfridge** [comment]
If I could go back in time, I would change `EqF` to correspond with `OrdF`. Our `EqF` is (I think) not really used much, because you have to know that the type indices are the same at compile time, whereas `OrdF` is heterogeneous, and the result gives you information both at the type- and value-level. `EqF` seems like a natural name for the corresponding equality type class.

Of course, going back and auditing all uses of `TestEquality` in Galois's codebase would be a nightmare. But if I had magical powers, that's probably what I would do.

I am not a huge fan of making our own `TestEquality` that conflicts with base, because I think `TestEquality` is already a bad name for the way we use it, and if we're going to fix the problem we should put our big boy pants on and fix the name while we're at it.

**Comment by robdockins** [comment]
I agree, `EqF` isn't terribly useful as-is. I'd be in favor of repurposing it, unless it is widely used in some way I'm not aware of.

Fortunately, we can make this transition gradually if we decide to do it, since `TestEquality` isn't going anywhere and we can have side-by-side instances for a time.

**Comment by benjaminselfridge** [comment]
I think repurposing `EqF` for this seems like a good idea; the only concern I have is that I'm not sure if `EqF` is actually used anywhere. But I'd be willing to spend a few hours investigating this.

**Comment by robdockins** [comment]
A moment of greping around leads me to believe we'll find very few uses.  The only "real" use of `eqF` I can find is in the `Eq` instance for `Pair`, which flows into the `Eq` instance for `MapF`.

**Comment by eddywestbrook** [comment]
FWIW, the Heapster code uses `testEquality` a bunch for testing heterogeneous equality of expressions. As does the Hobbits library. If we have to call that typeclass something else, then that's fine, it's just annoying... :(

**Comment by benjaminselfridge** [comment]
I think Rob's point probably applies -- you should be able to keep using `testEquality` for as long as you feel like using it, and the migration to `eqF` (if it happens at all) can be gradual, I think.
